src/ol/renderer/canvas/Layer.js

```diff
@@ -284,6 +285,20 @@ class CanvasLayerRenderer extends LayerRenderer {
     );
     const context = this.context;
 
+    const layer = this.getLayer();
+    const layerExtent = layer.getExtent();
+    if (layerExtent) {
+      const renderCoordinate = applyTransform(
+        frameState.pixelToCoordinateTransform,
+        pixel.slice()
+      );
+
+      /** get only data inside of the layer extent */
+      if (!containsXY(layerExtent, renderCoordinate[0], renderCoordinate[1])) {
```

Comment:
You could also use `containsCoordinate(layerExtent, renderCoordinate)` here.

Reply:
@ahocevar thanks for the hint, have missed that this function is available.

I also have created a temporal branch on my repo with an [example](https://github.com/boeckMt/openlayers/blob/canvas-layer-renderer-test/examples/hitdetect-image.js) so it is easier to discuss the Issue.

What is supposed to happen when there are layers without extent and one layer runs in the catch block of  [getDataAtPixel](https://github.com/boeckMt/openlayers/blob/canvas-layer-renderer-test/src/ol/renderer/canvas/Layer.js#L318)?
I think then it is not possible to detect this layer, because there is no extent and  `context.getImageData` is forbidden.

Would it then be fine to do the following check and don't return directly from the catch block.
```js
...
} catch (err) {
      if (err.name === 'SecurityError') {
        // tainted canvas, we assume there is data at the given pixel (although there might not be)
        // return new Uint8Array(); // do not return here
        data = new Uint8Array();
      }
      // return data;  // if there is an error getting the getImageData, pass it further to the next check
    }

    if (data[3] === 0 || !data.length) {
      return null;
    }
return data;
...
```
Or should `getDataAtPixel` return the empty array and then the check for data length should be happen in `forEachLayerAtPixel` like you suggested at the beginning.

> if (data && data.length) {
> https://github.com/openlayers/openlayers/issues/10099#issuecomment-725463303

Reply:
I don't think we need to change anything in addition. If we have a layer with a tainted canvas, our best guess (when this fix is merged) is to assume there is data when we hit the layer inside its extent. When there is no extent, we assume that there is data everywhere on the layer, like the inline comment says.

Reply:
That's fine, I further added two tests for this fix. Hope the usage of the openlayers Image https://openlayers.org/assets/theme/img/logo70.png for testing cross-data is ok.